File: python/shield/exchange.py
"""
Shield Key Exchange - Key exchange without public-key crypto.

Methods:
1. PAKE: Password-Authenticated Key Exchange
2. Physical: QR codes, base64 for manual exchange
3. Key Splitting: Threshold secret sharing

Example:
    >>> from shield.exchange import PAKEExchange, KeySplitter
    >>>
    >>> # Password-based key exchange
    >>> alice_key = PAKEExchange.derive("shared_password", b"salt", "alice")
    >>> bob_key = PAKEExchange.derive("shared_password", b"salt", "bob")
    >>> session_key = PAKEExchange.combine(alice_key, bob_key)
    >>>
    >>> # Key splitting
    >>> shares = KeySplitter.split(secret_key, num_shares=5)
    >>> recovered = KeySplitter.combine(shares)  # Need all shares
"""

import hmac
import hashlib
import struct
import secrets
import base64
from typing import List, Optional, Tuple


class PAKEExchange:
    """
    Password-Authenticated Key Exchange.

    Both parties derive a shared key from a common password.
    Uses role binding to prevent reflection attacks.
    """

    ITERATIONS = 200000  # Higher than normal for key exchange

    @staticmethod
    def derive(
        password: str,
        salt: bytes,
        role: str,
        iterations: int = None,
    ) -> bytes:
        """
        Derive key contribution from password.

        Args:
            password: Shared password between parties
            salt: Public salt (can be exchanged openly)
            role: Role identifier ('alice', 'bob', 'initiator', etc.)
            iterations: PBKDF2 iterations (default: 200000)

        Returns:
            32-byte key contribution
        """
        if iterations is None:
            iterations = PAKEExchange.ITERATIONS

        base_key = hashlib.pbkdf2_hmac(
            'sha256',
            password.encode(),
            salt,
            iterations
        )
        return hashlib.sha256(base_key + role.encode()).digest()

    @staticmethod
    def combine(*contributions: bytes) -> bytes:
        """
        Combine key contributions into session key.

        Args:
            *contributions: Key contributions from all parties

        Returns:
            32-byte shared session key
        """
        combined = b''.join(sorted(contributions))
        return hashlib.sha256(combined).digest()

    @staticmethod
    def generate_salt() -> bytes:
        """Generate random salt for key exchange."""
        return secrets.token_bytes(16)


class QRExchange:
    """
    Key exchange via QR codes or manual transfer.

    Encodes keys in URL-safe base64 for easy scanning/typing.
    """

    @staticmethod
    def encode(key: bytes) -> str:
        """
        Encode key for QR code or manual transfer.

        Args:
            key: Key bytes to encode

        Returns:
            URL-safe base64 string
        """
        return base64.urlsafe_b64encode(key).decode('ascii')

    @staticmethod
    def decode(encoded: str) -> bytes:
        """
        Decode key from QR code or manual input.

        Args:
            encoded: Base64 string from encode()

        Returns:
            Key bytes
        """
        return base64.urlsafe_b64decode(encoded)

    @staticmethod
    def generate_exchange_data(key: bytes, metadata: dict = None) -> str:
        """
        Generate complete exchange data with optional metadata.

        Args:
            key: Key to exchange
            metadata: Optional metadata (issuer, expiry, etc.)

        Returns:
            JSON-like string for QR code
        """
        import json
        data = {
            'v': 1,  # Version
            'k': base64.urlsafe_b64encode(key).decode('ascii'),
        }
        if metadata:
            data['m'] = metadata
        return json.dumps(data, separators=(',', ':'))

    @staticmethod
    def parse_exchange_data(data: str) -> Tuple[bytes, Optional[dict]]:
        """
        Parse exchange data from QR code.

        Args:
            data: JSON string from generate_exchange_data()

        Returns:
            Tuple of (key, metadata)
        """
        import json
        parsed = json.loads(data)
        key = base64.urlsafe_b64decode(parsed['k'])
        metadata = parsed.get('m')
        return key, metadata


class KeySplitter:
    """
    Split keys into shares for threshold recovery.

    This is a simplified XOR-based scheme where ALL shares
    are required for reconstruction. For true threshold
    schemes (k-of-n), use Shamir's Secret Sharing.
    """

    @staticmethod
    def split(key: bytes, num_shares: int) -> List[bytes]:
        """
        Split key into shares (all required for reconstruction).

        Args:
            key: Key to split
            num_shares: Number of shares to create

        Returns:
            List of shares
        """
        if num_shares < 2:
            raise ValueError("Need at least 2 shares")

        shares = [secrets.token_bytes(len(key)) for _ in range(num_shares - 1)]

        # Final share = XOR of key with all other shares
        final = key
        for share in shares:
            final = bytes(a ^ b for a, b in zip(final, share))

        shares.append(final)
        return shares

    @staticmethod
    def combine(shares: List[bytes]) -> bytes:
        """
        Combine shares to recover key.

        Args:
            shares: All shares from split()

        Returns:
            Original key
        """
        if len(shares) < 2:
            raise ValueError("Need at least 2 shares")

        result = shares[0]
        for share in shares[1:]:
            result = bytes(a ^ b for a, b in zip(result, share))

        return result

    @staticmethod
    def split_threshold(key: bytes, threshold: int, num_shares: int) -> List[Tuple[int, bytes]]:
        """
        Split key with threshold recovery (simplified Shamir-like).

        Uses polynomial interpolation simulation with XOR.
        Note: This is NOT cryptographically equivalent to Shamir's
        but provides basic threshold functionality.

        Args:
            key: Key to split
            threshold: Minimum shares needed
            num_shares: Total shares to create

        Returns:
            List of (index, share) tuples
        """
        if threshold > num_shares:
            raise ValueError("Threshold cannot exceed num_shares")
        if threshold < 2:
            raise ValueError("Threshold must be at least 2")

        # Generate random coefficients for "polynomial"
        coefficients = [key] + [secrets.token_bytes(len(key)) for _ in range(threshold - 1)]

        shares = []
        for i in range(1, num_shares + 1):
            # Evaluate "polynomial" at point i
            share = bytes(len(key))
            for j, coef in enumerate(coefficients):
                # Simplified: XOR with coefficient * i^j (modular)
                factor = pow(i, j, 256)
                term = bytes((c * factor) % 256 for c in coef)
                share = bytes(a ^ b for a, b in zip(share, term))
            shares.append((i, share))

        return shares

    @staticmethod
    def combine_threshold(shares: List[Tuple[int, bytes]], threshold: int) -> bytes:
        """
        Combine threshold shares to recover key.

        Args:
            shares: At least `threshold` shares from split_threshold()
            threshold: Threshold used in split

        Returns:
            Original key
        """
        if len(shares) < threshold:
            raise ValueError(f"Need at least {threshold} shares, got {len(shares)}")

        # Use first `threshold` shares
        used_shares = shares[:threshold]

        # Lagrange interpolation at x=0 (simplified)
        key_len = len(used_shares[0][1])
        result = bytes(key_len)

        for i, (xi, yi) in enumerate(used_shares):
            # Calculate Lagrange basis polynomial at 0
            numerator = 1
            denominator = 1
            for j, (xj, _) in enumerate(used_shares):
                if i != j:
                    numerator = (numerator * (-xj)) % 256
                    denominator = (denominator * (xi - xj)) % 256

            # Modular inverse (simplified)
            if denominator == 0:
                denominator = 1
            coef = (numerator * pow(denominator, -1, 256)) % 256

            term = bytes((y * coef) % 256 for y in yi)
            result = bytes(a ^ b for a, b in zip(result, term))

        return result

File: csharp/Shield/Exchange.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace Guard8.Shield
{
    /// <summary>
    /// Key Exchange - Key exchange without public-key crypto.
    ///
    /// Methods:
    /// 1. PAKE: Password-Authenticated Key Exchange
    /// 2. QR: QR codes, base64 for manual exchange
    /// 3. Key Splitting: XOR-based secret sharing
    /// </summary>
    public static class Exchange
    {
        /// <summary>
        /// Password-Authenticated Key Exchange.
        ///
        /// Both parties derive a shared key from a common password.
        /// Uses role binding to prevent reflection attacks.
        /// </summary>
        public static class PAKE
        {
            public const int DefaultIterations = 200000;

            /// <summary>
            /// Derive key contribution from password.
            /// </summary>
            /// <param name="password">Shared password between parties</param>
            /// <param name="salt">Public salt (can be exchanged openly)</param>
            /// <param name="role">Role identifier ('alice', 'bob', 'initiator', etc.)</param>
            /// <param name="iterations">PBKDF2 iterations (default: 200000)</param>
            /// <returns>32-byte key contribution</returns>
            public static byte[] Derive(string password, byte[] salt, string role, int iterations = 200000)
            {
                using var pbkdf2 = new Rfc2898DeriveBytes(password, salt, iterations, HashAlgorithmName.SHA256);
                byte[] baseKey = pbkdf2.GetBytes(32);

                using var sha256 = SHA256.Create();
                byte[] roleBytes = Encoding.UTF8.GetBytes(role);
                byte[] combined = baseKey.Concat(roleBytes).ToArray();
                return sha256.ComputeHash(combined);
            }

            /// <summary>
            /// Combine key contributions into session key.
            /// </summary>
            /// <param name="contributions">Key contributions from all parties</param>
            /// <returns>32-byte shared session key</returns>
            public static byte[] Combine(params byte[][] contributions)
            {
                // Sort contributions for deterministic output
                var sorted = contributions.OrderBy(b => Convert.ToHexString(b)).ToArray();

                using var sha256 = SHA256.Create();
                byte[] combined = sorted.SelectMany(b => b).ToArray();
                return sha256.ComputeHash(combined);
            }

            /// <summary>
            /// Generate random salt for key exchange.
            /// </summary>
            public static byte[] GenerateSalt()
            {
                return Shield.RandomBytes(16);
            }
        }

        /// <summary>
        /// Key exchange via QR codes or manual transfer.
        ///
        /// Encodes keys in URL-safe base64 for easy scanning/typing.
        /// </summary>
        public static class QR
        {
            /// <summary>
            /// Encode key for QR code or manual transfer.
            /// </summary>
            /// <param name="key">Key bytes to encode</param>
            /// <returns>URL-safe base64 string</returns>
            public static string Encode(byte[] key)
            {
                return Convert.ToBase64String(key).Replace('+', '-').Replace('/', '_').TrimEnd('=');
            }

            /// <summary>
            /// Decode key from QR code or manual input.
            /// </summary>
            /// <param name="encoded">Base64 string from Encode()</param>
            /// <returns>Key bytes</returns>
            public static byte[] Decode(string encoded)
            {
                // Restore standard base64 chars and padding
                string base64 = encoded.Replace('-', '+').Replace('_', '/');
                switch (base64.Length % 4)
                {
                    case 2: base64 += "=="; break;
                    case 3: base64 += "="; break;
                }
                return Convert.FromBase64String(base64);
            }

            /// <summary>
            /// Generate complete exchange data with optional metadata.
            /// </summary>
            /// <param name="key">Key to exchange</param>
            /// <param name="metadata">Optional metadata (issuer, expiry, etc.)</param>
            /// <returns>JSON-like string for QR code</returns>
            public static string GenerateExchangeData(byte[] key, Dictionary<string, object> metadata = null)
            {
                var data = new Dictionary<string, object>
                {
                    ["v"] = 1,
                    ["k"] = Encode(key)
                };

                if (metadata != null && metadata.Count > 0)
                {
                    data["m"] = metadata;
                }

                return JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = false });
            }

            /// <summary>
            /// Parse exchange data from QR code.
            /// </summary>
            /// <param name="data">JSON string from GenerateExchangeData()</param>
            /// <returns>Tuple of (key, metadata)</returns>
            public static (byte[] Key, Dictionary<string, object> Metadata) ParseExchangeData(string data)
            {
                var doc = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(data);
                string keyB64 = doc["k"].GetString();
                byte[] key = Decode(keyB64);

                Dictionary<string, object> metadata = null;
                if (doc.ContainsKey("m"))
                {
                    metadata = new Dictionary<string, object>();
                    foreach (var prop in doc["m"].EnumerateObject())
                    {
                        metadata[prop.Name] = prop.Value.ValueKind switch
                        {
                            JsonValueKind.String => prop.Value.GetString(),
                            JsonValueKind.Number => prop.Value.GetInt64(),
                            JsonValueKind.True => true,
                            JsonValueKind.False => false,
                            _ => prop.Value.ToString()
                        };
                    }
                }

                return (key, metadata);
            }
        }

        /// <summary>
        /// Split keys into shares for threshold recovery.
        ///
        /// This is a simplified XOR-based scheme where ALL shares
        /// are required for reconstruction.
        /// </summary>
        public static class KeySplitter
        {
            /// <summary>
            /// Split key into shares (all required for reconstruction).
            /// </summary>
            /// <param name="key">Key to split</param>
            /// <param name="numShares">Number of shares to create</param>
            /// <returns>List of shares</returns>
            public static List<byte[]> Split(byte[] key, int numShares)
            {
                if (numShares < 2)
                    throw new ArgumentException("Need at least 2 shares");

                var shares = new List<byte[]>();

                // Generate random shares for all but the last
                for (int i = 0; i < numShares - 1; i++)
                {
                    shares.Add(Shield.RandomBytes(key.Length));
                }

                // Final share = XOR of key with all other shares
                byte[] finalShare = (byte[])key.Clone();
                foreach (var share in shares)
                {
                    for (int i = 0; i < finalShare.Length; i++)
                    {
                        finalShare[i] ^= share[i];
                    }
                }
                shares.Add(finalShare);

                return shares;
            }

            /// <summary>
            /// Combine shares to recover key.
            /// </summary>
            /// <param name="shares">All shares from Split()</param>
            /// <returns>Original key</returns>
            public static byte[] Combine(List<byte[]> shares)
            {
                if (shares.Count < 2)
                    throw new ArgumentException("Need at least 2 shares");

                byte[] result = (byte[])shares[0].Clone();
                for (int i = 1; i < shares.Count; i++)
                {
                    for (int j = 0; j < result.Length; j++)
                    {
                        result[j] ^= shares[i][j];
                    }
                }

                return result;
            }
        }
    }
}
